File: Python/simulation/camera.py
class Camera:
    def __init__(self, sim_area_width, sim_area_height):
        self.zoom_level = 1.0
        self.pan_offset = [sim_area_width / 2, sim_area_height / 2]
        self.is_panning = False
        self.pan_start_screen = (0, 0)
        self.sim_area_width = sim_area_width
        self.sim_area_height = sim_area_height

    def world_to_screen(self, world_pos):
        x = world_pos[0] * self.zoom_level + self.pan_offset[0]
        y = world_pos[1] * self.zoom_level + self.pan_offset[1]
        return int(x), int(y)

    def screen_to_world(self, screen_pos):
        if self.zoom_level == 0:
            return 0, 0
        x = (screen_pos[0] - self.pan_offset[0]) / self.zoom_level
        y = (screen_pos[1] - self.pan_offset[1]) / self.zoom_level
        return x, y

    def start_pan(self, screen_pos):
        self.is_panning = True
        self.pan_start_screen = screen_pos

    def update_pan(self, screen_pos):
        if not self.is_panning:
            return
        dx = screen_pos[0] - self.pan_start_screen[0]
        dy = screen_pos[1] - self.pan_start_screen[1]
        self.pan_offset[0] += dx
        self.pan_offset[1] += dy
        self.pan_start_screen = screen_pos

    def end_pan(self):
        self.is_panning = False

    def zoom(self, factor, mouse_pos=None):
        old_zoom = self.zoom_level
        self.zoom_level = min(3.0, max(0.2, self.zoom_level * factor))
        if mouse_pos:
            world_pos = self.screen_to_world(mouse_pos)
            self.pan_offset[0] = mouse_pos[0] - world_pos[0] * self.zoom_level
            self.pan_offset[1] = mouse_pos[1] - world_pos[1] * self.zoom_level
        else:
            center_screen = (self.sim_area_width / 2, self.sim_area_height / 2)
            center_world = self.screen_to_world(center_screen)
            self.pan_offset[0] = center_screen[0] - center_world[0] * self.zoom_level
            self.pan_offset[1] = center_screen[1] - center_world[1] * self.zoom_level

    def reset_view(self):
        self.zoom_level = 1.0
        self.pan_offset = [self.sim_area_width / 2, self.sim_area_height / 2]

File: Website/js/camera.js
class Camera {
    constructor(screenWidth, screenHeight) {
        this.screenWidth = screenWidth;
        this.screenHeight = screenHeight;
        this.panX = 0;
        this.panY = 0;
        this.zoomLevel = 1.0;
        this.minZoom = 0.2;
        this.maxZoom = 5.0;
        this.zoomStep = 0.1;
        this.isPanning = false;
        this.lastMousePos = null;
    }

    // Convert world coordinates to screen coordinates
    worldToScreen(worldPos) {
        const screenX = (worldPos[0] * this.zoomLevel) + this.panX;
        const screenY = (worldPos[1] * this.zoomLevel) + this.panY;
        return [screenX, screenY];
    }

    // Convert screen coordinates to world coordinates
    screenToWorld(screenPos) {
        const worldX = (screenPos[0] - this.panX) / this.zoomLevel;
        const worldY = (screenPos[1] - this.panY) / this.zoomLevel;
        return [worldX, worldY];
    }

    // Start panning
    startPan(mousePos) {
        this.isPanning = true;
        this.lastMousePos = mousePos;
    }

    // Update panning
    pan(mousePos) {
        if (!this.isPanning || !this.lastMousePos) return;
        
        const deltaX = mousePos[0] - this.lastMousePos[0];
        const deltaY = mousePos[1] - this.lastMousePos[1];
        
        this.panX += deltaX;
        this.panY += deltaY;
        
        this.lastMousePos = mousePos;
    }

    // Stop panning
    stopPan() {
        this.isPanning = false;
        this.lastMousePos = null;
    }

    // Zoom at the specific screen position
    zoomAt(delta, centerScreenPos) {
        const centerWorldBefore = this.screenToWorld(centerScreenPos);
        
        // Apply zoom
        if (delta > 0) {
            this.zoomLevel = Math.min(this.maxZoom, this.zoomLevel + this.zoomStep);
        } else {
            this.zoomLevel = Math.max(this.minZoom, this.zoomLevel - this.zoomStep);
        }
        
        // Adjust pan to keep the point under the mouse in the same position
        const centerWorldAfter = this.screenToWorld(centerScreenPos);
        this.panX += (centerWorldAfter[0] - centerWorldBefore[0]) * this.zoomLevel;
        this.panY += (centerWorldAfter[1] - centerWorldBefore[1]) * this.zoomLevel;
    }

    // Reset the camera to initial state
    reset() {
        this.panX = this.screenWidth / 2;
        this.panY = this.screenHeight / 2;
        this.zoomLevel = 1.0;
    }
} 